contracts/CollectiveGovernance.sol

```diff
@@ -151,62 +135,194 @@ contract CollectiveGovernance is Governance, VoteStrategy {
+        uint256[] memory _shareList = _class.discover(msg.sender);
+        uint256 count = 0;
+        for (uint256 i = 0; i < _shareList.length; i++) {
+            uint256 shareId = _shareList[i];
+            count += _storage.voteForByShare(_proposalId, msg.sender, shareId);
+        }
+        if (count > 0) {
+            emit VoteTally(_proposalId, msg.sender, count);
+        } else {
+            revert("Not voter");
+        }
+    }
+
+    function voteForWithTokenId(uint256 _proposalId, uint256 _tokenId)
         public
-        requireStrategyVersion(_proposalId)
-        requireVoter(_proposalId, msg.sender)
         requireVoteOpen(_proposalId)
+        requireNoVeto(_proposalId)
     {
-        _storage._castVoteFor(_proposalId, msg.sender);
+        uint256 count = _storage.voteForByShare(_proposalId, msg.sender, _tokenId);
+        if (count > 0) {
+            emit VoteTally(_proposalId, msg.sender, count);
+        } else {
+            revert("Not voter");
+        }
+    }
+
+    function voteForWithTokenList(uint256 _proposalId, uint256[] memory _tokenIdList)
+        external
+        requireVoteOpen(_proposalId)
+        requireNoVeto(_proposalId)
+    {
+        uint256 count = 0;
+        for (uint256 i = 0; i < _tokenIdList.length; i++) {
+            count += _storage.voteForByShare(_proposalId, msg.sender, _tokenIdList[i]);
+        }
+        if (count > 0) {
+            emit VoteTally(_proposalId, msg.sender, count);
+        } else {
+            revert("Not voter");
+        }
     }
 
     // @notice undo any previous vote
-    function undoVote(uint256 _proposalId)
+    function undoVote(uint256 _proposalId) public requireVoteOpen(_proposalId) requireNoVeto(_proposalId) {
+        VoterClass _class = _storage.voterClass(_proposalId);
+        uint256[] memory _shareList = _class.discover(msg.sender);
+        uint256 count = 0;
+        for (uint256 i = 0; i < _shareList.length; i++) {
+            uint256 shareId = _shareList[i];
+            count += _storage.undoVoteById(_proposalId, msg.sender, shareId);
+        }
+        if (count > 0) {
+            emit VoteUndo(_proposalId, msg.sender, count);
+        } else {
+            revert("Not voter");
+        }
+    }
+
+    function undoWithTokenId(uint256 _proposalId, uint256 _tokenId)
         public
-        requireStrategyVersion(_proposalId)
-        requireVoter(_proposalId, msg.sender)
         requireVoteOpen(_proposalId)
+        requireNoVeto(_proposalId)
     {
-        _storage._castVoteUndo(_proposalId, msg.sender);
+        uint256 count = _storage.undoVoteById(_proposalId, msg.sender, _tokenId);
+        if (count > 0) {
+            emit VoteUndo(_proposalId, msg.sender, count);
+        } else {
+            revert("Not voter");
+        }
     }
 
-    function voteAgainst(uint256 _proposalId)
+    function voteAgainst(uint256 _proposalId) public requireVoteOpen(_proposalId) requireNoVeto(_proposalId) {
```

Comment:
Can we consolidate voteFor, voteAgainst and voteAbstain into 1 function where the voter just sends the option that they want? This would also allow us to easily abstract to arbitrary vote options set by the proposal creator like snapshot has

Reply:
That would be an architectural change.   As it stands these methods are simple clear and easy to understand.   This will be good from both a literate code, security and gas point of view.   In the future we will probably have more diversity int he voting strategies.